### DISCUSSION TITLE
Question about the detailled router capabilities

### DISCUSSION THREAD

**Discussion by lustefan97** [The-OpenROAD-Project/OpenROAD#3963] [Q&A]
Hello,

So I want to implement a Face-to-face 3D flow on OpenROAD. I found that there is the define_pin_shape_pattern function and the documentation says that it could be used to help implement exactly that. 

What I did was simply take a small design to begin with (ibex) and place all of its I/O pins on the top layer (so a little more than 250, not a lot), to see what would happen. I tried with both the Skywater PDK and the Asap7 PDK, and in both cases the detailled routing step never converged towards an optimal solution without violations. One thing to know is that I already had a design to route with ONE pin on the top layer, even if the design was a very large and sparely placed ibex core. I haven't tried an "in between" case yet.

I think I noticed (just an observation, I didn't properly tested that yet) that the number of violation increased a lot in "3D" (pins on the top layer) compared to the 2D case. To give an idea, using the ASAP7 PDK, the number of violations in 2D at the beginning of the optimisation iteration number one (so the second iteration) was around 6000, while in the 3D case it shot up to 22000 violations. 
The behaviour of the number of violation in 2D was to decrease a lot, maybe then increase a little bit, but go to zero after while, in 3D, it started to "oscillate" around a value (that value seems to depend on the placement density and die size).

I have let the flow run for a few hours in the 3D case (compared to maybe 20 minutes in the 2D case) without the number of violations ever really coming down (especially for such a small design, it seems strange). Usually in 2D the flow stopped after a few optimisation iterations, but in 3D even after 20 iterations (or 50+ with the Skywater PDK) it was still trying.

My questions are thus :

1. Has the detailled routing ever finished in one of your test when multiples IO pins were on the top layer ? If so, was there a specific way that you used to setup the flow, for example reducing routing resources beforehand to prevent some eventual congestion due to the wires going to the I/O pins ?
2. Can the tool become "overwhelmed" by the number of violations, and start running in circles while trying to handle them ?
3. More generally, what are the limits of TritonRoute ?

**Comment by maliberty** [comment]
We have routed to top level bumps.  This would be better as an issue with a test case showing the problem as this is too vague to say much.

**Reply by lustefan97** [reply]
I'll do that ! About the test case, are the script and setup files that I used enough ?

**Reply by maliberty** [reply]
The easiest in ORFS is to do 'make detail_route_issue' which will package everything needed.

**Reply by lustefan97** [reply]
I unfortunately am not using the automated flow just yet, I pilot the flow in the gui using TCL scripts while in _OpenROAD/test_ directory. I added a directory Test3D in which I have everything I use to test things. I feel like it is more flexible. Will that directory be enough ?

**Reply by maliberty** [reply]
So long as I can unpack it and reproduce the problem with a single invocation of OR that should be fine.

**Reply by lustefan97** [reply]
Thank you, I will create an issue with the compressed directory so as to reproduce it
